<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="This package allows to query multiple taxonomic names on the NCBI and BOLD repositories and retrieve DNA barcodes and associated metadata for any wanted marker. It heavily relies on the bold and rentrez packages from the rOpenSci, and takes advantages of the parallelizing framework from the future package to speed up the retrieval of data from the NCBI, respecting their API requests limit. The final output is a tibble object modified following the formatting requirements of the refdb package, and it is cleaned from mining duplicates, differences in formats of BOLD and NCBI metadata (e.g. geographic coordinates, dates) and commonly occurring issues that derive from downloading data from those repositories. In synthesis, it provides a unified framework for mining DNA Barcodes from the main repositories, roviding clean, metadata-rich sequences in a programmatic way or interactively, depending on the wanted usage.">
<title>Fast and reliable mining of DNA Barcodes from the NCBI and BOLD â€¢ barcodeMineR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Fast and reliable mining of DNA Barcodes from the NCBI and BOLD">
<meta property="og:description" content="This package allows to query multiple taxonomic names on the NCBI and BOLD repositories and retrieve DNA barcodes and associated metadata for any wanted marker. It heavily relies on the bold and rentrez packages from the rOpenSci, and takes advantages of the parallelizing framework from the future package to speed up the retrieval of data from the NCBI, respecting their API requests limit. The final output is a tibble object modified following the formatting requirements of the refdb package, and it is cleaned from mining duplicates, differences in formats of BOLD and NCBI metadata (e.g. geographic coordinates, dates) and commonly occurring issues that derive from downloading data from those repositories. In synthesis, it provides a unified framework for mining DNA Barcodes from the main repositories, roviding clean, metadata-rich sequences in a programmatic way or interactively, depending on the wanted usage.">
<meta property="og:image" content="https://matteoce.github.io/barcodeMineR/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">barcodeMineR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="articles/api_rate_barcodeMineR.html">Speeding up the recovery of DNA barcodes</a>
    <a class="dropdown-item" href="articles/intro_barcodeMineR.html">Introduction to the barcodeMineR package</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/MatteoCe/barcodeMineR/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><p><strong>NOTE: This is a temporary GitHub repository created for testing purposes.</strong></p>
<div class="section level1">
<div class="page-header">
<img src="logo.png" class="logo" alt=""><h1 id="barcodeminer-">barcodeMineR <a class="anchor" aria-label="anchor" href="#barcodeminer-"></a>
</h1>
</div>
<!-- badges: start -->

<div class="section level2">
<h2 id="the-barcodeminer-package">The barcodeMineR package<a class="anchor" aria-label="anchor" href="#the-barcodeminer-package"></a>
</h2>
<p>This package allows to query multiple taxonomic names on the NCBI and BOLD repositories and retrieve DNA barcodes and associated metadata for any wanted marker. It heavily relies on the <a href="https://github.com/ropensci/bold" class="external-link">bold</a> and <a href="https://github.com/ropensci" class="external-link">rentrez</a> packages from the <a href="https://ropensci.org/" class="external-link">rOpenSci</a>, and takes advantage of the asynchronous framework from the <a href="https://github.com/HenrikBengtsson/future" class="external-link">future</a> package to speed up the retrieval of data from the NCBI, respecting its API requests rate limit.</p>
<p>The final output is a data frame object modified following the formatting requirements of the <a href="https://github.com/fkeck/refdb" class="external-link">refdb</a> package, and it is cleaned from mining duplicates, differences in formats of BOLD and NCBI metadata (<em>e.g.</em> geographic coordinates, dates) and commonly occurring issues that derive from downloading and merging data from those two repositories.</p>
<p>In synthesis, it provides a unified framework for mining DNA Barcodes from the main online genomic repositories, providing clean, metadata-rich sequences in a programmatic way or interactively, depending on the wanted usage.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>As the package is still in development, you can install this version of barcodeMineR, directly from the GitHub repository, using the <a href="https://www.bioconductor.org/install/" class="external-link">Bioconductor</a> package:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"MatteoCe/barcodeMineR"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="basic-usage">Basic usage<a class="anchor" aria-label="anchor" href="#basic-usage"></a>
</h2>
<p>The basic functioning of the package includes only two steps, which must be run consequently:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MatteoCe/barcodeMineR" class="external-link">barcodeMineR</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># check if a species is on the NCBI nucleotide database:</span></span>
<span><span class="va">tax</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/get_ncbi_taxonomy.html">get_ncbi_taxonomy</a></span><span class="op">(</span><span class="st">"Dissostichus mawsoni"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># download all records of this species from the database:</span></span>
<span><span class="va">rec</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/download_ncbi.html">download_ncbi</a></span><span class="op">(</span><span class="va">tax</span>, ask <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># display output:</span></span>
<span><span class="va">rec</span></span>
<span><span class="co">#&gt; # A tibble: 189 Ã— 30</span></span>
<span><span class="co">#&gt;    recordID   markerCode DNA_seq  phylum class order family genus species source</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;      &lt;chr&gt;      &lt;DNA&gt;    &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt; </span></span>
<span><span class="co">#&gt;  1 HM422302.1 COI        CTCTACTâ€¦ Chordâ€¦ Actiâ€¦ Percâ€¦ Nototâ€¦ Dissâ€¦ Dissosâ€¦ NCBI  </span></span>
<span><span class="co">#&gt;  2 ON000293.1 COX1       GCGCCTGâ€¦ Chordâ€¦ Actiâ€¦ Percâ€¦ Nototâ€¦ Dissâ€¦ Dissosâ€¦ NCBI  </span></span>
<span><span class="co">#&gt;  3 MK843765.1 COI        TCTCTACâ€¦ Chordâ€¦ Actiâ€¦ Percâ€¦ Nototâ€¦ Dissâ€¦ Dissosâ€¦ NCBI  </span></span>
<span><span class="co">#&gt;  4 DQ498816.1 Cytb       GCCACCCâ€¦ Chordâ€¦ Actiâ€¦ Percâ€¦ Nototâ€¦ Dissâ€¦ Dissosâ€¦ NCBI  </span></span>
<span><span class="co">#&gt;  5 DQ498794.1 Rhod       GCCTACAâ€¦ Chordâ€¦ Actiâ€¦ Percâ€¦ Nototâ€¦ Dissâ€¦ Dissosâ€¦ NCBI  </span></span>
<span><span class="co">#&gt;  6 MK500763.1 enc1       TCTGACGâ€¦ Chordâ€¦ Actiâ€¦ Percâ€¦ Nototâ€¦ Dissâ€¦ Dissosâ€¦ NCBI  </span></span>
<span><span class="co">#&gt;  7 MG729451.1 COI        GAACTTAâ€¦ Chordâ€¦ Actiâ€¦ Percâ€¦ Nototâ€¦ Dissâ€¦ Dissosâ€¦ NCBI  </span></span>
<span><span class="co">#&gt;  8 KY656477.1 COI        GCCGGAAâ€¦ Chordâ€¦ Actiâ€¦ Percâ€¦ Nototâ€¦ Dissâ€¦ Dissosâ€¦ NCBI  </span></span>
<span><span class="co">#&gt;  9 LC138011.1 ND1        ATGCTTTâ€¦ Chordâ€¦ Actiâ€¦ Percâ€¦ Nototâ€¦ Dissâ€¦ Dissosâ€¦ NCBI  </span></span>
<span><span class="co">#&gt; 10 LC138011.1 ND2        ATGAGCCâ€¦ Chordâ€¦ Actiâ€¦ Percâ€¦ Nototâ€¦ Dissâ€¦ Dissosâ€¦ NCBI  </span></span>
<span><span class="co">#&gt; # â„¹ 179 more rows</span></span>
<span><span class="co">#&gt; # â„¹ 20 more variables: lat &lt;dbl&gt;, lon &lt;dbl&gt;, lengthGene &lt;int&gt;, sampleID &lt;chr&gt;,</span></span>
<span><span class="co">#&gt; #   QueryName &lt;chr&gt;, identified_by &lt;chr&gt;, taxNotes &lt;lgl&gt;, db_xref &lt;chr&gt;,</span></span>
<span><span class="co">#&gt; #   sourceID &lt;chr&gt;, NCBI_ID &lt;chr&gt;, institutionStoring &lt;lgl&gt;,</span></span>
<span><span class="co">#&gt; #   collected_by &lt;chr&gt;, collection_date &lt;chr&gt;, altitude &lt;lgl&gt;, depth &lt;lgl&gt;,</span></span>
<span><span class="co">#&gt; #   country &lt;chr&gt;, directionPrimers &lt;chr&gt;, lengthSource &lt;int&gt;,</span></span>
<span><span class="co">#&gt; #   PCR_primers &lt;chr&gt;, note &lt;chr&gt;</span></span></code></pre></div>
<p>Take a look at the <a href="https://matteoce.github.io/barcodeMineR/articles/intro_barcodeMineR.html">introductory vignette</a> for a more complete tutorial!</p>
<p>After that, proceed with the <a href="https://matteoce.github.io/barcodeMineR/articles/api_rate_barcodeMineR.html">â€˜Speeding upâ€™ vignette</a> to learn how to increase speed and reliability of the package functions.</p>
</div>
</div>

  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/MatteoCe/barcodeMineR/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/MatteoCe/barcodeMineR/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing barcodeMineR</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Matteo Cecchetto <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-4505-4104" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/MatteoCe/barcodeMineR/actions/workflows/R-CMD-check.yaml" class="external-link"><img src="https://github.com/MatteoCe/barcodeMineR/actions/workflows/R-CMD-check.yaml/badge.svg" alt="R-CMD-check"></a></li>
<li><a href="https://github.com/MatteoCe/barcodeMineR/actions/workflows/test-coverage.yaml" class="external-link"><img src="https://github.com/MatteoCe/barcodeMineR/actions/workflows/test-coverage.yaml/badge.svg" alt="test-coverage"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Matteo Cecchetto.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
