---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

**NOTE: This is a temporary GitHub repository created for testing purposes.**

# barcodeMineR <img src="man/figures/barcodeMineR_logo.png" align="right" height="139" alt="" />

<!-- badges: start -->
[![R-CMD-check](https://github.com/MatteoCe/barcodeMineR/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/MatteoCe/barcodeMineR/actions/workflows/R-CMD-check.yaml)
[![test-coverage](https://github.com/MatteoCe/barcodeMineR/actions/workflows/test-coverage.yaml/badge.svg)](https://github.com/MatteoCe/barcodeMineR/actions/workflows/test-coverage.yaml)
<!-- badges: end -->

## The barcodeMineR package

This package allows to query multiple taxonomic names on the NCBI and BOLD repositories and retrieve DNA barcodes and associated metadata for any wanted marker. It heavily relies on the [bold](https://github.com/ropensci/bold) and [rentrez](https://github.com/ropensci) packages from the [rOpenSci](https://ropensci.org/), and takes advantage of the asynchronous framework from the [future](https://github.com/HenrikBengtsson/future) package to speed up the retrieval of data from the NCBI, respecting its API requests rate limit.

The final output is a data frame object modified following the formatting requirements of the [refdb](https://github.com/fkeck/refdb) package, and it is cleaned from mining duplicates, differences in formats of BOLD and NCBI metadata (*e.g.* geographic coordinates, dates) and commonly occurring issues that derive from downloading and merging data from those two repositories.

In synthesis, it provides a unified framework for mining DNA Barcodes from the main online genomic repositories, providing clean, metadata-rich sequences in a programmatic way or interactively, depending on the wanted usage.

## Installation

As the package is still in development, you can install this version of barcodeMineR, directly from the GitHub repository, using the [Bioconductor](https://www.bioconductor.org/install/) package:

``` r
BiocManager::install("MatteoCe/barcodeMineR")
```

## Basic usage

The basic functioning of the package includes only two steps, which must be run consequently:

```{r example}
library(barcodeMineR)

# check if a species is on the NCBI nucleotide database:
tax <- get_ncbi_taxonomy("Dissostichus mawsoni")

# download all records of this species from the database:
rec <- download_ncbi(tax, ask = FALSE)

# display output:
rec

```

Take a look at the [introductory vignette](https://matteoce.github.io/barcodeMineR/articles/intro_barcodeMineR.html) for a more complete tutorial!

After that, proceed with the ['Speeding up' vignette](https://matteoce.github.io/barcodeMineR/articles/api_rate_barcodeMineR.html) to learn how to increase speed and reliability of the package functions.

