---
title: "Introduction to the barcodeMineR package"
output: rmarkdown::html_vignette
description: >
  Learn how to use the barcodeMineR package
vignette: >
  %\VignetteIndexEntry{Introduction to the barcodeMineR package}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```
```{r setup}
library(barcodeMineR)
```

**NOTE: This package is still in development and awaiting for CRAN approval.**

The _barcodeMineR_ package allows to recover DNA barcodes from a vector of taxonomic names.
The most basic usage consists in the recovery of the taxonomic information first, followed by the download of all records available. The function [get_ncbi_taxonomy](https://matteoce.github.io/barcodeMineR/reference/get_ncbi_taxonomy.html) does exactly that:

```{r}
library(barcodeMineR)

# search taxonomic information for a species on the NCBI
tax <- get_ncbi_taxonomy("Dissostichus mawsoni")
tax
```

Taxonomic information for the species _Dissostichus mawsoni_ is available at the NCBI taxonomy database, and we obtained a data frame class object with basic information regarding it.

This object can be used for the next command, which will retrieve all DNA barcodes available corresponding to this species at the NCBI nucleotide database. One of the main features of the next function consists in separating the different CDS and rRNA products of each accession number. You can leave the argument _ask_ to default (TRUE) and wait to be asked to choose which CDS/rRNA to select. For now, we leave this to FALSE, so we retain automatically all CDS/rRNA features:

```{r}
rec_NCBI <- download_ncbi(tax, ask = FALSE)

```

Now we have the final _refdb_ formatted data frame including all DNA barcodes available at the NCBI for the species _Dissostichus mawsoni_.

```{r, echo = FALSE}
rec_NCBI
```

This data frame includes `r nrow(rec_NCBI)` records (rows). However, if we were to search for all _Dissostichus mawsoni_ records on the NCBI nucleotide database, using the same default filters called inside the _barcodeMineR_ functions, we will notice that there are less accession numbers than the number of records obtained by the [download_ncbi](https://matteoce.github.io/barcodeMineR/reference/download_ncbi.html) function:

```{r}
# the default filters exclude whole genome shotgun sequences and transcribed 
# shotgun assembly products
rentrez::entrez_search(db="nucleotide", term="(((txid36200[ORGN] NOT wgs[Keyword]) NOT tsa[Keyword]) AND biomol_genomic[PROP]) AND (cds[Feature key] OR rrna[Feature key])")
```

The package recovered every single DNA barcode/gene/product available for the accession numbers corresponding
to _Dissostichus mawsoni_, thus, for example, for the accession number 'LC138011.1' we obtained both ND1 and ND2 markers as separate records in the _refdb_ data frame.

In case you want to choose which marker to keep in the final tibble, set the _ask_ argument to TRUE (default), and, when asked, choose which marker to download data for:

```{r}
rec_NCBI <- download_ncbi(tax)
```

The DNA barcodes will be in their 5'-to-3' direction and of the correct length defined by the [INSDC location descriptor](https://www.insdc.org/submitting-standards/feature-table/#3.4.3), as they've been extracted from the single fasta sequence of the same accession number.

The same procedure can be applied using the functions querying the BOLD database, which have the same naming style. Here, we're retrieving the taxonomy information from the BOLD database for the same species:

```{r}
# search taxonomic information for a species on BOLD
tax <- get_bold_taxonomy("Dissostichus mawsoni")
tax

```

The output is slightly different from the output of [get_ncbi_taxonomy](https://matteoce.github.io/barcodeMineR/reference/get_ncbi_taxonomy.html), however, it serves the same role. In fact, it can then be used to download all _Dissostichus mawsoni_ records from the BOLD database.

```{r}
rec_BOLD <- download_bold(tax, ask = FALSE)
rec_BOLD
```

We have retrieved `r nrow(rec_BOLD)` records from the BOLD database, corresponding to `r length(unique(rec_BOLD$recordID))` different process-IDs. BOLD hosts other markers, but the vast majority of records correspond to Cytochrome c oxidase I sequences.

Searching species names, or taxa in general, using the [taxonomy functions](https://matteoce.github.io/barcodeMineR/reference/index.html#taxonomy) follows a set of rules that is described in the vignette "[Searching taxonomy](https://matteoce.github.io/barcodeMineR/articles/taxonomy_barcodeMineR.html)".

## Merging data from different repositories

We can now merge both _refdb_ data frames to obtain a single one. The default behaviour of the next function will check which records have been mined from the other repository and return only the original record. If needed, this behaviour can be switched off by setting the argument _resolve.conflicts_ to FALSE:

```{r}
total <- mergeBarcodeOres(rec_NCBI, rec_BOLD)
```

As we can see from the printed messages, some records were actually mined versions of other records obtained from the other repository. They have been removed after checking that the original version was downloaded from the opposite repository.

```{r}
total
```

This is it, now we have all DNA barcodes available on the major online repositories for the species _Dissostichus mawsoni_. 

In case we want to add additional, private sequences to the mined results, we can use the function [loadBarcodeOre](https://matteoce.github.io/barcodeMineR/reference/loadBarcodeOre.html) and specify either the paths to a tsv file and a fasta file including our new sequences, or to a data frame and a _DNAStringSet_ object with the same, to output a _refdb_ data frame formatted as those we merged earlier. 

In order to do that, the tsv file must include the same fields (*i.e.* column names) included in the example dataset object _example_record_.

```{r, eval = FALSE}
# load example datasets from barcodeMineR...
rec <- barcodeMineR::example_record
seq <- barcodeMineR::example_sequence

#... or include the path to your files
rec <- "~/home/my/path/data.tsv"
seq <- "~/home/my/path/data.fasta"

new_records <- loadBarcodeOre(rec, seq)
new_records
```
```{r, echo = FALSE}
rec <- barcodeMineR::example_record
seq <- barcodeMineR::example_sequence
new_records <- loadBarcodeOre(rec, seq)
new_records
```

We can merge these records with the ones we mined online and examine all results:

```{r}
total <- mergeBarcodeOres(total, new_records)
total
```

## Analyzing the database

Having a pre-formatted [refdb](https://fkeck.github.io/refdb/index.html) data frame allows to use the functions of that package, including filtering operations, quality checks, examination of taxonomic conflicts and plotting functions. For example, if we were to search for taxonomic conflicts, we could use the appropriate function from the _refdb_ package:

```{r}
library(refdb)

refdb_check_tax_conflict(total)
```

As observed from the output, it appears that the nomenclature for ray-finned fishes is different in the BOLD and NCBI taxonomies (Actinopteri vs Actinopterygii).

We can also filter records based on the length of the sequences. Although we retain this information in the lengthGene field, the _refdb_ package directly calculates it using the [refdb_filter_seq_length](https://fkeck.github.io/refdb/reference/refdb_filter_seq_length.html) function. Let's say we want to know how many records correspond to a maximum of 658 base pairs (bp) in length:

```{r}
total_filt <- refdb_filter_seq_length(total, max_len = 658)
```

A total of `r nrow(total_filt)` records include sequences up to 658 bp in length. However, many records results much shorter than this length:

```{r, message = FALSE}
refdb_plot_seqlen_hist(total_filt)
```

In fact, many records have been extracted from complete mitochondrial genomes, and this can be readily inspected by using one of the plotting functions of the _barcodeMineR_ package:

```{r}
plot_length(total_filt)
```

As shown, the original length of the fasta sequence of some records is much longer than the final result obtained with the _barcodeMineR_ package.

Not much information was retrieved from each record, but we can see how many records include the primers used to amplify their sequence with another plotting function of the _barcodeMineR_ package, credit by [Gwiazdowski et al. (2015)](https://journals.plos.org/plosone/article/file?id=10.1371/journal.pone.0125635&type=printable):

```{r}
plot_primers(total_filt)
```

We can also map the records with available geographical coordinates using the _refdb_ package:

```{r, message = FALSE, warning = FALSE}
refdb_plot_map(total)
```

Only a few records have geographical coordinates in the final merged table.

## Exporting the results

Once one or more _refdb_ objects has been obtained, these can be exported as any data frame. As an example, here it is shown how to export one as a TSV file, using the package _readr_:

```{r, eval = FALSE}
readr::write_delim(x = total, file = "~/my/path/to/file.tsv", delim = "\t")
```

Likewise, if this file needs to be loaded as a simple data.frame object, the same method can be applied. However, in case this object needs to be loaded as a _refdb_ data frame, then the fields set by the _barcodeMineR_ functions need to be preserved in a _yaml_ file designed specifically for its outputs. For this reason, when saving a _refdb_ object obtained with the _barcodeMineR_ functions, a _yaml_ file must be stored alongside it. This can be done using the _refdb_ package:

```{r, eval = FALSE}
# save the fields settings in a yml file format
refdb_write_fields(total, "~/my/path/to/fields.yml")

# load the tsv saved before and set the fields 
df <- readr::read_delim(file="~/my/path/to/file.tsv", delim = "\t")
ref <- refdb_set_fields(df, config_yaml = "~/my/path/to/fields.yml")
```

This is the basic usage of the __barcodeMineR__ package. Now you can check how to speed up the recovery of records from the NCBI database by reading this [vignette](https://matteoce.github.io/barcodeMineR/articles/api_rate_barcodeMineR.html).

