<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This function searches for NCBI records corresponding to the species found in
the argument `ncbi_tax`, i.e. the output from the function `get_ncbi_taxonomy`.
It can also search for accession numbers and return the same type of object.
For a thorough explanation of the function usage and capabilities, see the
'Introduction to the barcodeMineR package' vignette:
vignette(&quot;Introduction to the barcodeMineR package&quot;, package = &quot;barcodeMineR&quot;)"><title>Download records from the NCBI — download_ncbi • barcodeMineR</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Download records from the NCBI — download_ncbi"><meta property="og:description" content="This function searches for NCBI records corresponding to the species found in
the argument `ncbi_tax`, i.e. the output from the function `get_ncbi_taxonomy`.
It can also search for accession numbers and return the same type of object.
For a thorough explanation of the function usage and capabilities, see the
'Introduction to the barcodeMineR package' vignette:
vignette(&quot;Introduction to the barcodeMineR package&quot;, package = &quot;barcodeMineR&quot;)"><meta property="og:image" content="https://matteoce.github.io/barcodeMineR/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">barcodeMineR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/index.html">Articles</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/MatteoCe/barcodeMineR/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Download records from the NCBI</h1>
      <small class="dont-index">Source: <a href="https://github.com/MatteoCe/barcodeMineR/blob/HEAD/R/download_ncbi.R" class="external-link"><code>R/download_ncbi.R</code></a></small>
      <div class="d-none name"><code>download_ncbi.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function searches for NCBI records corresponding to the species found in
the argument `ncbi_tax`, i.e. the output from the function `get_ncbi_taxonomy`.</p>
<p>It can also search for accession numbers and return the same type of object.</p>
<p>For a thorough explanation of the function usage and capabilities, see the
'Introduction to the barcodeMineR package' vignette:
<code>vignette("Introduction to the barcodeMineR package", package = "barcodeMineR")</code></p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">download_ncbi</span><span class="op">(</span></span>
<span>  ncbi_tax <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ncbi_ids <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  rate_xml <span class="op">=</span> <span class="fl">200</span>,</span>
<span>  rate_fasta <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  default.filter <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  filter <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  api_rate <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ask <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  prefix <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>ncbi_tax</dt>
<dd><p>`data.frame` A data frame, as returned from the
`get_ncbi_taxonomy()` function.</p></dd>


<dt>ncbi_ids</dt>
<dd><p>`character` A character vector with NCBI accession numbers.</p></dd>


<dt>rate_xml</dt>
<dd><p>`integer` The number of xml objects to be downloaded at a
time. It can be lowered for unstable internet connections. Defaults to
`200`.</p></dd>


<dt>rate_fasta</dt>
<dd><p>`integer` The number of fasta sequences to be downloaded
at a time. Many fasta can correspond to mitogenomes and chromosomes, which
may lead to an errors if downloaded in great numbers. Defaults to `100`.</p></dd>


<dt>default.filter</dt>
<dd><p>`logical` Whether to filter the records excluding whole
genome shotgun sequences and transcribed shotgun assembly. Defaults to
`TRUE`.</p></dd>


<dt>filter</dt>
<dd><p>`character` an additional query filter in the form of
a/multiple string/s to add to every searched taxid. This will allow any
user to specifically filter every search with a custom query. Multiple
strings should be provided in the form of a character vector of single
query filters (see description for details). Defaults to `NULL`.</p></dd>


<dt>api_rate</dt>
<dd><p>`integer` The API rate with which to iterate each separate
request. Must be a number between 3 and 10 which will translate in a rate
of `1 / api_rate` seconds.</p></dd>


<dt>ask</dt>
<dd><p>`logical` Should the function ask the user whether to filter the
final output for taxonomic ranks. Default `TRUE`.</p></dd>


<dt>prefix</dt>
<dd><p>`character` A character string that will be used to create
numbered custom ids for each record in ascending order. The prefix will
compose the recordID field in the final object. Default to `NULL`, using
the internal recordID generator that will use the accession number for NCBI
records and the processID for BOLD records, avoiding duplicates by adding
`_1`, `_2` etc.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>`data.frame` A refdb data frame, including the DNA sequence as a
  field.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>[download_bold()]</p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">tax</span> <span class="op">&lt;-</span> <span class="fu"><a href="get_ncbi_taxonomy.html">get_ncbi_taxonomy</a></span><span class="op">(</span><span class="st">"Polymastia invaginata"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">download_ncbi</span><span class="op">(</span><span class="va">tax</span>, ask <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 16 × 30</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    recordID   markerCode  DNA_seq phylum class order family genus species source</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;DNA&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> HG423800.1 28S rRNA    <span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #BB0000; font-weight: bold;">A</span><span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #00BB00; font-weight: bold;">CCC</span>… Porif… Demo… Poly… Polym… Poly… Polyma… NCBI  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> HG423799.1 28S rRNA    <span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #BB0000; font-weight: bold;">A</span><span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #00BB00; font-weight: bold;">CCC</span>… Porif… Demo… Poly… Polym… Poly… Polyma… NCBI  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> HG423770.1 28S rRNA    <span style="color: #BB0000; font-weight: bold;">A</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #BB0000; font-weight: bold;">A</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #BBBB00; font-weight: bold;">GG</span>… Porif… Demo… Poly… Polym… Poly… Polyma… NCBI  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> HG423769.1 28S rRNA    <span style="color: #BB0000; font-weight: bold;">A</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #BB0000; font-weight: bold;">A</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #BBBB00; font-weight: bold;">GG</span>… Porif… Demo… Poly… Polym… Poly… Polyma… NCBI  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> HG423740.1 28S rRNA    <span style="color: #0000BB; font-weight: bold;">TT</span><span style="color: #BB0000; font-weight: bold;">AA</span><span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #00BB00; font-weight: bold;">C</span>… Porif… Demo… Poly… Polym… Poly… Polyma… NCBI  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> HG423739.1 28S rRNA    <span style="color: #0000BB; font-weight: bold;">TT</span><span style="color: #BB0000; font-weight: bold;">AA</span><span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #00BB00; font-weight: bold;">C</span>… Porif… Demo… Poly… Polym… Poly… Polyma… NCBI  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> HG423712.1 COI         <span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #BB0000; font-weight: bold;">A</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">T</span>… Porif… Demo… Poly… Polym… Poly… Polyma… NCBI  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> HG423711.1 COI         <span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #BB0000; font-weight: bold;">A</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">T</span>… Porif… Demo… Poly… Polym… Poly… Polyma… NCBI  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> LN850239.1 COI         <span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #BB0000; font-weight: bold;">A</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #0000BB; font-weight: bold;">T</span>… Porif… Demo… Poly… Polym… Poly… Polyma… NCBI  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> LN850219.1 COI         <span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #BB0000; font-weight: bold;">A</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #0000BB; font-weight: bold;">T</span>… Porif… Demo… Poly… Polym… Poly… Polyma… NCBI  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">11</span> LN850218.1 COI         <span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #BB0000; font-weight: bold;">A</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #0000BB; font-weight: bold;">T</span>… Porif… Demo… Poly… Polym… Poly… Polyma… NCBI  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">12</span> LN606560.1 28S rRNA    <span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #BB0000; font-weight: bold;">A</span><span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #00BB00; font-weight: bold;">CCC</span>… Porif… Demo… Poly… Polym… Poly… Polyma… NCBI  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">13</span> LN606530.1 28S rRNA    <span style="color: #BB0000; font-weight: bold;">A</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #BB0000; font-weight: bold;">A</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #BBBB00; font-weight: bold;">GG</span>… Porif… Demo… Poly… Polym… Poly… Polyma… NCBI  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">14</span> LN606500.1 28S rRNA    <span style="color: #0000BB; font-weight: bold;">TT</span><span style="color: #BB0000; font-weight: bold;">AA</span><span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #00BB00; font-weight: bold;">C</span>… Porif… Demo… Poly… Polym… Poly… Polyma… NCBI  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">15</span> LN606462.1 COI         <span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #BB0000; font-weight: bold;">A</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">T</span>… Porif… Demo… Poly… Polym… Poly… Polyma… NCBI  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">16</span> AY561922.1 28S large … <span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #BBBB00; font-weight: bold;">GG</span><span style="color: #00BB00; font-weight: bold;">CCC</span>… Porif… Demo… Poly… Polym… Poly… Polyma… NCBI  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 20 more variables: lat &lt;dbl&gt;, lon &lt;dbl&gt;, lengthGene &lt;int&gt;, sampleID &lt;chr&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   QueryName &lt;chr&gt;, identified_by &lt;lgl&gt;, taxNotes &lt;lgl&gt;, db_xref &lt;chr&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   sourceID &lt;chr&gt;, NCBI_ID &lt;chr&gt;, institutionStoring &lt;lgl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   collected_by &lt;lgl&gt;, collection_date &lt;chr&gt;, altitude &lt;lgl&gt;, depth &lt;lgl&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   country &lt;lgl&gt;, directionPrimers &lt;lgl&gt;, lengthSource &lt;int&gt;,</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;">#   PCR_primers &lt;lgl&gt;, note &lt;lgl&gt;</span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Matteo Cecchetto.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer></div>

  

  

  </body></html>

