<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Learn how to improve speed and reliability
">
<title>Speeding up the recovery of DNA barcodes • barcodeMineR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Speeding up the recovery of DNA barcodes">
<meta property="og:description" content="Learn how to improve speed and reliability
">
<meta property="og:image" content="https://matteoce.github.io/barcodeMineR/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">barcodeMineR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../articles/index.html">Articles</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/MatteoCe/barcodeMineR/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Speeding up the recovery of DNA barcodes</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/MatteoCe/barcodeMineR/blob/HEAD/vignettes/api_rate_barcodeMineR.Rmd" class="external-link"><code>vignettes/api_rate_barcodeMineR.Rmd</code></a></small>
      <div class="d-none name"><code>api_rate_barcodeMineR.Rmd</code></div>
    </div>

    
    
<p><strong>NOTE: This is a temporary GitHub repository created for
testing purposes.</strong></p>
<p>The time required to mine DNA barcodes from the NCBI nucleotide
database depends on how many species/taxa are searched and how many
records correspond to each taxa. Due to the structure of the
<strong>barcodeMineR</strong> functions, and its consequent usage,
different approaches can be adopted to speed up the procedures, with
varying effects.</p>
<p>Although the rationale and functioning behind each method is
described below, including the <a href="https://matteoce.github.io/barcodeMineR/articles/api_rate_barcodeMineR.html#tuning-the-functions-arguments">fine
tuning</a> of the functions’ arguments, it is suggested to follow the
functions’ default parameters, and, at the same time, adopt all of the
approaches here described, which do not only speed up the recovery of
DNA barcodes, but also reduce excessive blocking from the servers.</p>
<div class="section level2">
<h2 id="api-rate-limit">API rate limit<a class="anchor" aria-label="anchor" href="#api-rate-limit"></a>
</h2>
<p>The NCBI servers allows up to only three requests per second from a
single user. If this limit is not respected, the NCBI will block further
requests for some time. If a request takes less than 1/3 of a second to
finish, then this represents the maximum rate that can be adopted for
these analyses.</p>
<p>For example, if we have a vector of 100 species, we can expect a
theoretical execution time for the <a href="https://matteoce.github.io/barcodeMineR/reference/get_ncbi_taxonomy.html">get_ncbi_taxonomy</a>
function of 66.67 seconds:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MatteoCe/barcodeMineR" class="external-link">barcodeMineR</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># extract example dataset with 200 species from the Ross Sea (Antarctica, Southern Ocean)</span></span>
<span><span class="va">specs</span> <span class="op">&lt;-</span> <span class="fu">barcodeMineR</span><span class="fu">::</span><span class="va"><a href="../reference/species200.html">species200</a></span></span>
<span></span>
<span><span class="va">tax</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_ncbi_taxonomy.html">get_ncbi_taxonomy</a></span><span class="op">(</span><span class="va">specs</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">]</span>, ask <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>And, during the rendering of this vignettes, it executed in 1.54
minutes.</p>
<p>This api rate, which is the default for the <em>ncbi</em> functions
of the <em>barcodeMineR</em> package, is, however, highly subjected to
changes in the speed of the internet connections, and might delay some
requests at certain times, leading to an accumulation of four requests
in a window of one second of time.</p>
<p>In order to avoid this, a higher download rate can be adopted if the
user decides to register a <a href="https://account.ncbi.nlm.nih.gov/" class="external-link">NCBI account</a>. Using an API
key, the user can send up to 10 requests per second to the NCBI.</p>
<p>This can be done “externally”, in respect to the
<em>barcodeMineR</em> functions, following the <a href="https://cran.r-project.org/web/packages/rentrez/vignettes/rentrez_tutorial.html#rate-limiting-and-api-keys" class="external-link">rentrez
documentation</a>, specifically with this function:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set_entrez_key</span><span class="op">(</span><span class="st">"ABCD123"</span><span class="op">)</span></span></code></pre></div>
<p>By setting an NCBI account, the user assures that requests that don’t
take much time can be executed even faster, and avoid excessive blocking
when the internet connection is particularly unreliable. Including an
NCBI API key approximately halves the execution time of the <a href="https://matteoce.github.io/barcodeMineR/reference/get_ncbi_taxonomy.html">get_ncbi_taxonomy</a>,
when querying multiple species at once.</p>
<p>The above command is executed, during the rendering of this vignette,
in 44 seconds.</p>
<p>All ncbi functions of the <em>barcodeMineR</em> package would benefit
from the inclusion of an API key, which increases speed of fast requests
and reliability at long execution times.</p>
</div>
<div class="section level2">
<h2 id="adopting-an-asynchronous-framework">Adopting an asynchronous framework<a class="anchor" aria-label="anchor" href="#adopting-an-asynchronous-framework"></a>
</h2>
<p>One of the main advantages of the <strong>barcodeMineR</strong>
package consists in the capacity to adopt the asynchronous parallelizing
framework of the <a href="https://github.com/HenrikBengtsson/future" class="external-link">future</a> package to
speed up the download of records from the NCBI repository.</p>
<p>The execution time of the <a href="https://matteoce.github.io/barcodeMineR/reference/download_ncbi.html">download_ncbi</a>
function depends on how many records correspond to each single
species/taxa on the NCBI nucleotide database.</p>
<p>For example, if we wanted to search for all records corresponding to
the Antarctic krill species <em>Euphausia superba</em>, we could expect
to download 1584 records, after an execution time of approximately 1.44
minutes, for the function <em>download_ncbi</em>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tax</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_ncbi_taxonomy.html">get_ncbi_taxonomy</a></span><span class="op">(</span><span class="st">"Euphausia superba"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># the execution time refers to the function below only</span></span>
<span><span class="va">recs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/download_ncbi.html">download_ncbi</a></span><span class="op">(</span><span class="va">tax</span>, ask <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>In this context, the asynchronous parallelizing framework of the <a href="https://future.futureverse.org/" class="external-link">future</a> package comes in
handy. In synthesis, each request is sent as a background process, which
will not block the current R session to send another request at a
specified time, provided that there are enough “workers” (cores)
available. This means that, using the <em>future</em> framework, we can
send a request every 1/3 of a second (or 1/10, if we set the NCBI API
key), independently from the completion of the previous request.</p>
<p>We can setup the <em>future</em> framework using its <a href="https://future.futureverse.org/articles/future-1-overview.html" class="external-link">functions</a>.
Here, we’re setting up the “multisession” plan using all available cores
from our machine:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">future</span><span class="fu">::</span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="st">"multisession"</span><span class="op">)</span></span></code></pre></div>
<p>You can specify the number of cores using the <em>workers</em>
argument, and check how many cores are available using the following
command:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">future</span><span class="fu">::</span><span class="fu"><a href="https://parallelly.futureverse.org/reference/availableCores.html" class="external-link">availableCores</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; system </span></span>
<span><span class="co">#&gt;      4</span></span></code></pre></div>
<p>We can revert to the “normal”, sequential framework when we’re done
with our work using the <em>barcodeMineR</em> package, supplying
“sequential” to the same command. Remember to do this before proceeding
with other, unrelated work, in case you’re not interested in using the
same asynchronous framework. If you’re unsure of which plan the current
R session is set to, use the same command without arguments.</p>
<p>This plan (“multisession”) works on all OSes, on both Rstudio or the
command line. It will significantly speed up the recovery of records
from the NCBI, especially for the <em>download_ncbi</em> function, but
can also improve the speed of <em>get_ncbi_taxonomy</em> (see <a href="https://matteoce.github.io/barcodeMineR/articles/api_rate_barcodeMineR.html#tuning-the-functions-arguments">this
section</a> for more details).</p>
<p>Now, the entire process of downloading <em>Euphausia superba</em>
records will take sensibly less time than in “sequential”
<em>future</em> plan, providing the final results after 1.01
minutes.</p>
<div class="section level3">
<h3 id="tips">Tips<a class="anchor" aria-label="anchor" href="#tips"></a>
</h3>
<ul>
<li>Progress bar</li>
</ul>
<p>When downloading large numbers of records from BOLD or the NCBI
databases, it is suggested to include a progress bar. Again, here we’re
relying on another beautiful package, the <a href="https://github.com/HenrikBengtsson/progressr" class="external-link">progressr</a>
package. As for the <em>future</em> package, we don’t need to supply
specific arguments to the <em>barcodeMineR</em> functions to activate
it, but we can do it “externally”:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">progressr</span><span class="fu">::</span><span class="fu"><a href="https://progressr.futureverse.org/reference/handlers.html" class="external-link">handlers</a></span><span class="op">(</span>global<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu">progressr</span><span class="fu">::</span><span class="fu"><a href="https://progressr.futureverse.org/reference/handlers.html" class="external-link">handlers</a></span><span class="op">(</span><span class="st">"progress"</span><span class="op">)</span></span></code></pre></div>
<p>Now, a progress bar will appear while downloading records, and
different messages will occasionally be printed to inform the user of
the current status of the operation. The progress bar can be modified,
according to the user’s preferences, as reported in the <a href="https://cran.r-project.org/web/packages/progressr/vignettes/progressr-intro.html" class="external-link">progressr
documentation</a>.</p>
<ul>
<li>Multisession and <em>get_ncbi_taxonomy</em>:</li>
</ul>
<p>Adopting the future asynchronous framework can significantly improve
the execution of the <em>download_ncbi</em> function, but can also
improve the speed of <em>get_ncbi_taxonomy</em>, when the “multisession”
plan is set with two to four workers (cores). In fact, setting up a
“future”, a background process, takes the <em>future</em> package some
time, and, for requests that do not take much time it is suggested to
not include too many workers in a “multisession” plan, as they would not
significantly improve the speed.</p>
<p>Below, the execution time of the function <em>get_ncbi_taxonomy</em>,
at different <em>future</em> plans is shown, revealing how most of the
improvement is observed after including the NCBI API key:
<img src="api_rate_barcodeMineR_files/figure-html/unnamed-chunk-12-1.png" width="700">
## Why can’t I speed up the BOLD functions?</p>
<p>BOLD servers maintenance <a href="https://github.com/ropensci/bold/issues/76#issuecomment-789909664" class="external-link">does
not support an API rate limit</a>, like the NCBI does, and, for this
reason, a certain number of requests sent in a particular window of time
may result in a temporary blocking by the BOLD servers. The factors the
mostly influence the occurrence of this inconvenience are two:</p>
<ul>
<li>Number of requests:</li>
</ul>
<p>The bold functions have default settings that try to avoid sending
too many requests at a time. For example, the default <em>api_rate</em>
argument setting of <a href="https://matteoce.github.io/barcodeMineR/reference/get_bold_taxonomy.html">get_bold_taxonomy</a>
allows to search a specific taxon every 1/0.06 seconds, meaning that the
maximum number of requests is set to 250 per hour. Although it can be
considered pretty low, this comes from recent testing (May 2024) that
detected frequent blocking when querying the BOLD taxonomy database when
surpassing this limit. This only applies to taxa that do not include
children, thus only for species. When searching higher level taxa
leaving the <em>descend</em> argument to TRUE (default), by relying on
the usage of the package <a href="https://github.com/ropensci/taxize" class="external-link">taxize</a>, the limit might be
more easily surpassed when many children are included.</p>
<ul>
<li>Size of requested data:</li>
</ul>
<p>Differently from the <em>get_bold_taxonomy</em> function, the <a href="https://matteoce.github.io/barcodeMineR/reference/download_bold.html">download_bold</a>
function has a higher limit, which mostly depends on the amount of data
that is requested each time. When querying a long list of species names,
the function is set to send each request every second, without frequent
blocking. However, this depends on the amount of data that correspond to
each taxon. If, for example, we were to search for all Arthropoda
sequences on BOLD, although the <em>get_bold_taxonomy</em> function
would run smoothly, provided that the <em>descend</em> argument is set
to FALSE, the <em>download_bold</em> would surely fail, resulting in a
temporary blocking by the BOLD servers. When using
<em>download_bold</em>, only search <a href="https://github.com/ropensci/bold/issues/29" class="external-link">low level
taxonomies</a> (downward from family), as suggested by the <a href="https://github.com/ropensci/bold?tab=readme-ov-file#large-data" class="external-link">bold
package</a>.</p>
</div>
</div>
<div class="section level2">
<h2 id="tuning-the-functions-arguments">Tuning the functions’ arguments<a class="anchor" aria-label="anchor" href="#tuning-the-functions-arguments"></a>
</h2>
<p>Different arguments can be tuned to improve speed and/or reliability
of requests to the NCBI servers. These include the <em>rate</em> of xml
or fasta downloads and the <em>API rate</em>:</p>
<ul>
<li><p>The <strong>api_rate</strong> argument allows to override the
default settings of the NCBI requests rate. The future plan framework
allows to send requests every 1 / <strong>api_rate</strong> seconds,
thus ensuring that no more that ‘api_rate’ requests will be sent in a
window framework of 1 second. However, due to fluctuations in the
internet connection, still more than <strong>api_rate</strong> number of
requests might arrive to the NCBI servers, causing errors. The function
can handle up to 5 consecutive errors per request, but too many errors
might block the whole process. The <strong>api_rate</strong> parameter
can be modified in order to slow down the requests sent per second. It
overrides the automatic selection of the optimal parameter (either 3 or
10) and accepts one decimal degree number between 1.0 and 10.0, so if
the internet connection is particularly bad, it can be set between 2.5
and 2.0, for example, in order to slow the number of requests per second
and reduce the possibility of errors.</p></li>
<li><p>The <strong>rate</strong> argument of the <a href="https://matteoce.github.io/barcodeMineR/reference/get_ncbi_taxonomy.html">get_ncbi_taxonomy</a>
and <a href="https://matteoce.github.io/barcodeMineR/reference/download_bold.html">download_bold</a>
and the arguments <strong>rate_xml</strong> and
<strong>rate_fasta</strong> for the <a href="https://matteoce.github.io/barcodeMineR/reference/download_ncbi.html">download_ncbi</a>
function allow to set the number of taxa/species that will be queried
with each request and has impacts on the memory usage of the package.
Although decreasing or increasing their value does not significantly
impact the speed, it is worth mentioning that, in case of low memory
availability, reducing the amount of xml or fasta sequences downloaded
at a time may improve the performance. Nonetheless, it is better to keep
the values to default settings, which have been tested on different
settings.</p></li>
</ul>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Matteo Cecchetto.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
