<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="barcodeMineR">
<title>Speeding up the recovery of DNA barcodes • barcodeMineR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Speeding up the recovery of DNA barcodes">
<meta property="og:description" content="barcodeMineR">
<meta property="og:image" content="https://matteoce.github.io/barcodeMineR/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">barcodeMineR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/api_rate_barcodeMineR.html">Speeding up the recovery of DNA barcodes</a>
    <a class="dropdown-item" href="../articles/intro_barcodeMineR.html">Introduction to the barcodeMineR package</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/MatteoCe/barcodeMineR/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Speeding up the recovery of DNA barcodes</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/MatteoCe/barcodeMineR/blob/HEAD/vignettes/api_rate_barcodeMineR.Rmd" class="external-link"><code>vignettes/api_rate_barcodeMineR.Rmd</code></a></small>
      <div class="d-none name"><code>api_rate_barcodeMineR.Rmd</code></div>
    </div>

    
    
<p><strong>NOTE: This is a temporary GitHub repository created for
testing purposes.</strong></p>
<div class="section level2">
<h2 id="rate-limit-and-parallelization">Rate limit and parallelization<a class="anchor" aria-label="anchor" href="#rate-limit-and-parallelization"></a>
</h2>
<p>One of the main advantages of the <strong>barcodeMineR</strong>
package consists in leveraging the asynchronous parallelizing framework
of the <a href="https://github.com/HenrikBengtsson/future" class="external-link">future</a>
package to speed up the download of sequences from the NCBI
repository.</p>
<p>The following commands will retrieve 1837 records, corresponding to
the species <em>Promachocrinus kerguelensis</em>, taking approximately
2.11778817574183 minutes.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MatteoCe/barcodeMineR" class="external-link">barcodeMineR</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">tax</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_ncbi_taxonomy.html">get_ncbi_taxonomy</a></span><span class="op">(</span><span class="st">"Promachocrinus kerguelensis"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">recs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/download_ncbi.html">download_ncbi</a></span><span class="op">(</span><span class="va">tax</span>, ask <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>The NCBI servers allows up to only three requests per second from a
single user. If this limit is not respected, the NCBI will block further
requests for some time. “Classic” parallelization is not a viable
option, considering that different requests may take significantly
different time (we may send a request for DNA barcodes with only one CDS
or rRNA, while other will download whole genomes). For the same reason,
running sequentially (like in a lapply or for loop with a waiting time
of 1/3 seconds) is highly subjected to the execution time of each
request.</p>
<p>In this context, the asynchronous parallelizing framework of the
<strong>future</strong> package comes in handy. In synthesis, each
request is sent as a background process, which will not block the
current R session to send another request at a specified time, provided
that there are enough “workers” (cores) available. This means that,
using the <strong>future</strong> framework, we can send a request every
1/3 of a second, independently from the completion of the previous
request.</p>
<p>We can setup the <strong>future</strong> framework using its <a href="https://future.futureverse.org/articles/future-1-overview.html" class="external-link">functions</a>.
Here, we’re setting up the “multisession” plan using all available cores
from our machine:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">future</span><span class="fu">::</span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="st">"multisession"</span><span class="op">)</span></span></code></pre></div>
<p>You can specify the number of cores using the <em>workers</em>
argument, and check how many cores are available using the following
command:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">parallelly</span><span class="fu">::</span><span class="fu"><a href="https://parallelly.futureverse.org/reference/availableWorkers.html" class="external-link">availableWorkers</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "localhost" "localhost" "localhost" "localhost"</span></span></code></pre></div>
<p>We can revert to the “normal”, sequential framework when we’re done
with our work using the <strong>barcodeMineR</strong> package, supplying
“sequential” to the same command. Remember to do this before proceeding
with other, unrelated work, in case you’re not interested in using the
same asynchronous framework. If you’re unsure of which plan the current
R session is set to, use the same command without arguments.</p>
<p>This plan (“multisession”) works on all OSes, on both Rstudio or the
command line. It will significantly speed up the recovery of records
from the NCBI, especially for the
<code>{r, echo = TRUE, eval = FALSE}download_ncbi</code> function.</p>
<p>However, before proceeding with an example, when downloading large
numbers of records from BOLD or the NCBI databases, it is suggested to
include a progress bar. Again, here we’re relying on another beautiful
package, the <a href="https://github.com/HenrikBengtsson/progressr" class="external-link">progressr</a>
package. As before, we don’t need to supply specific arguments to the
<strong>barcodeMineR</strong> functions to activate it, but we can do it
“externally”:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">progressr</span><span class="fu">::</span><span class="fu"><a href="https://progressr.futureverse.org/reference/handlers.html" class="external-link">handlers</a></span><span class="op">(</span>global<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu">progressr</span><span class="fu">::</span><span class="fu"><a href="https://progressr.futureverse.org/reference/handlers.html" class="external-link">handlers</a></span><span class="op">(</span><span class="st">"progress"</span><span class="op">)</span></span></code></pre></div>
<p>Now, a progress bar will appear while downloading records, and
different messages will occasionally be printed to inform the user of
the current status of the operation. The progress bar can be modified,
according to the user’s preferences, as reported in the <a href="https://cran.r-project.org/web/packages/progressr/vignettes/progressr-intro.html" class="external-link">progressr
documentation</a>.</p>
<p>Now, the entire process of downloading <em>Promachocrinus
kerguelensis</em> will take sensibly less time than in “sequential”
<strong>future</strong> plan, providing the final results after 54
seconds.</p>
</div>
<div class="section level2">
<h2 id="stay-safe-and-fast-with-the-ncbi-api-key">Stay safe and fast with the NCBI API key<a class="anchor" aria-label="anchor" href="#stay-safe-and-fast-with-the-ncbi-api-key"></a>
</h2>
<p>An additional parameter can be supplied “externally”, allowing to
increase the speed of the
<code>{r, echo = TRUE, eval = FALSE}get_ncbi_taxonomy</code> function
and avoid blocking from the NCBI servers for both functions. As
mentioned above, the NCBI allows a maximum of three requests per second,
a limit that is respected by the default settings of the
<strong>barcodeMineR</strong> package. However, a higher download rate
can be adopted if the user decides to register a <a href="https://account.ncbi.nlm.nih.gov/" class="external-link">NCBI account</a>. Using an API
key, the user can send up to 10 requests per second to the NCBI. This
can be done following the <a href="https://cran.r-project.org/web/packages/rentrez/vignettes/rentrez_tutorial.html#rate-limiting-and-api-keys" class="external-link">rentrez
documentation</a>, specifically with the following function:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set_entrez_key</span><span class="op">(</span><span class="st">"ABCD123"</span><span class="op">)</span></span></code></pre></div>
<p>By setting an NCBI account, the user assures that requests that don’t
take much time can be executed even faster, and avoid excessive blocking
when the internet connection is particularly unreliable. Including an
NCBI API key halves the execution time of the
<code>{r, echo = TRUE, eval = FALSE}get_ncbi_taxonomy</code>, when
querying multiple species at once.</p>
</div>
<div class="section level2">
<h2 id="tuning-the-functions-arguments">Tuning the functions’ arguments<a class="anchor" aria-label="anchor" href="#tuning-the-functions-arguments"></a>
</h2>
<p>Different arguments can be tuned to improve speed and/or reliability
of requests to the NCBI servers. These include the <strong>rate</strong>
of xml or fasta downloads and the <strong>API rate</strong>. The first
one is</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by The package maintainer.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
