<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Introduction to the barcodeMineR package • barcodeMineR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to the barcodeMineR package">
<meta name="description" content="Learn how to use the barcodeMineR package
">
<meta property="og:description" content="Learn how to use the barcodeMineR package
">
<meta property="og:image" content="https://matteoce.github.io/barcodeMineR/logo.png">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">barcodeMineR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item"><a class="nav-link" href="../articles/index.html">Articles</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/MatteoCe/barcodeMineR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">



<link href="intro_barcodeMineR_files/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet">
<script src="intro_barcodeMineR_files/htmlwidgets-1.6.4/htmlwidgets.js"></script><link href="intro_barcodeMineR_files/leaflet-1.3.1/leaflet.css" rel="stylesheet">
<script src="intro_barcodeMineR_files/leaflet-1.3.1/leaflet.js"></script><link href="intro_barcodeMineR_files/leafletfix-1.0.0/leafletfix.css" rel="stylesheet">
<script src="intro_barcodeMineR_files/proj4-2.6.2/proj4.min.js"></script><script src="intro_barcodeMineR_files/Proj4Leaflet-1.0.1/proj4leaflet.js"></script><link href="intro_barcodeMineR_files/rstudio_leaflet-1.3.1/rstudio_leaflet.css" rel="stylesheet">
<script src="intro_barcodeMineR_files/leaflet-binding-2.2.2/leaflet.js"></script><link href="intro_barcodeMineR_files/leaflet-markercluster-1.0.5/MarkerCluster.css" rel="stylesheet">
<link href="intro_barcodeMineR_files/leaflet-markercluster-1.0.5/MarkerCluster.Default.css" rel="stylesheet">
<script src="intro_barcodeMineR_files/leaflet-markercluster-1.0.5/leaflet.markercluster.js"></script><script src="intro_barcodeMineR_files/leaflet-markercluster-1.0.5/leaflet.markercluster.freezable.js"></script><script src="intro_barcodeMineR_files/leaflet-markercluster-1.0.5/leaflet.markercluster.layersupport.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Introduction to the barcodeMineR package</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/MatteoCe/barcodeMineR/blob/master/vignettes/intro_barcodeMineR.Rmd" class="external-link"><code>vignettes/intro_barcodeMineR.Rmd</code></a></small>
      <div class="d-none name"><code>intro_barcodeMineR.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MatteoCe/barcodeMineR" class="external-link">barcodeMineR</a></span><span class="op">)</span></span></code></pre></div>
<p><strong>NOTE: This package is still in development and awaiting for
CRAN approval.</strong></p>
<p>The <em>barcodeMineR</em> package allows to recover DNA barcodes from
a vector of taxonomic names. The most basic usage consists in the
recovery of the taxonomic information first, followed by the download of
all records available. The function <a href="https://matteoce.github.io/barcodeMineR/reference/get_ncbi_taxonomy.html">get_ncbi_taxonomy</a>
does exactly that:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MatteoCe/barcodeMineR" class="external-link">barcodeMineR</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># search taxonomic information for a species on the NCBI</span></span>
<span><span class="va">tax</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_ncbi_taxonomy.html">get_ncbi_taxonomy</a></span><span class="op">(</span><span class="st">"Dissostichus mawsoni"</span><span class="op">)</span></span>
<span><span class="va">tax</span></span>
<span><span class="co">#&gt;              queryName taxid    rank       scientificName   phylum       class</span></span>
<span><span class="co">#&gt; 1 Dissostichus mawsoni 36200 species Dissostichus mawsoni Chordata Actinopteri</span></span>
<span><span class="co">#&gt;         order        family        genus              species</span></span>
<span><span class="co">#&gt; 1 Perciformes Nototheniidae Dissostichus Dissostichus mawsoni</span></span></code></pre></div>
<p>Taxonomic information for the species <em>Dissostichus mawsoni</em>
is available at the NCBI taxonomy database, and we obtained a data frame
class object with basic information regarding it.</p>
<p>This object can be used for the next command, which will retrieve all
DNA barcodes available corresponding to this species at the NCBI
nucleotide database. One of the main features of the next function
consists in separating the different CDS and rRNA products of each
accession number. You can leave the argument <em>ask</em> to default
(TRUE) and wait to be asked to choose which CDS/rRNA to select. For now,
we leave this to FALSE, so we retain automatically all CDS/rRNA
features:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rec_NCBI</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/download_ncbi.html">download_ncbi</a></span><span class="op">(</span><span class="va">tax</span>, ask <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>Now we have the final <em>refdb</em> formatted data frame including
all DNA barcodes available at the NCBI for the species <em>Dissostichus
mawsoni</em>.</p>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 192 × 30</span></span></span>
<span><span class="co">#&gt;    recordID   markerCode DNA_seq  phylum class order family genus species source</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;DNA&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> HM422302.1 COI        <span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #BB0000; font-weight: bold;">A</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">T</span>… Chord… Acti… Perc… Notot… Diss… Dissos… NCBI  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> ON000293.1 COX1       <span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #00BB00; font-weight: bold;">CC</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #BBBB00; font-weight: bold;">G</span>… Chord… Acti… Perc… Notot… Diss… Dissos… NCBI  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> MK843765.1 COI        <span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #BB0000; font-weight: bold;">A</span><span style="color: #00BB00; font-weight: bold;">C</span>… Chord… Acti… Perc… Notot… Diss… Dissos… NCBI  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> DQ498816.1 Cytb       <span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #00BB00; font-weight: bold;">CC</span><span style="color: #BB0000; font-weight: bold;">A</span><span style="color: #00BB00; font-weight: bold;">CCC</span>… Chord… Acti… Perc… Notot… Diss… Dissos… NCBI  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> DQ498794.1 Rhod       <span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #00BB00; font-weight: bold;">CC</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #BB0000; font-weight: bold;">A</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #BB0000; font-weight: bold;">A</span>… Chord… Acti… Perc… Notot… Diss… Dissos… NCBI  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> MK500763.1 enc1       <span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #BB0000; font-weight: bold;">A</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #BBBB00; font-weight: bold;">G</span>… Chord… Acti… Perc… Notot… Diss… Dissos… NCBI  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> MG729451.1 COI        <span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #BB0000; font-weight: bold;">AA</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">TT</span><span style="color: #BB0000; font-weight: bold;">A</span>… Chord… Acti… Perc… Notot… Diss… Dissos… NCBI  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> KY656477.1 COI        <span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #00BB00; font-weight: bold;">CC</span><span style="color: #BBBB00; font-weight: bold;">GG</span><span style="color: #BB0000; font-weight: bold;">AA</span>… Chord… Acti… Perc… Notot… Diss… Dissos… NCBI  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> LC138011.1 ND1        <span style="color: #BB0000; font-weight: bold;">A</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">TTT</span>… Chord… Acti… Perc… Notot… Diss… Dissos… NCBI  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> LC138011.1 ND2        <span style="color: #BB0000; font-weight: bold;">A</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #BB0000; font-weight: bold;">A</span><span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #00BB00; font-weight: bold;">CC</span>… Chord… Acti… Perc… Notot… Diss… Dissos… NCBI  </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 182 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 20 more variables: lat &lt;dbl&gt;, lon &lt;dbl&gt;, lengthGene &lt;int&gt;, sampleID &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   QueryName &lt;chr&gt;, identified_by &lt;chr&gt;, taxNotes &lt;lgl&gt;, db_xref &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   sourceID &lt;chr&gt;, NCBI_ID &lt;chr&gt;, institutionStoring &lt;lgl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   collected_by &lt;chr&gt;, collection_date &lt;chr&gt;, altitude &lt;lgl&gt;, depth &lt;lgl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   country &lt;lgl&gt;, directionPrimers &lt;chr&gt;, lengthSource &lt;int&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   PCR_primers &lt;chr&gt;, note &lt;chr&gt;</span></span></span></code></pre>
<p>This data frame includes 192 records (rows). However, if we were to
search for all <em>Dissostichus mawsoni</em> records on the NCBI
nucleotide database, using the same default filters called inside the
<em>barcodeMineR</em> functions, we will notice that there are less
accession numbers than the number of records obtained by the <a href="https://matteoce.github.io/barcodeMineR/reference/download_ncbi.html">download_ncbi</a>
function:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># the default filters exclude whole genome shotgun sequences and transcribed </span></span>
<span><span class="co"># shotgun assembly products</span></span>
<span><span class="fu">rentrez</span><span class="fu">::</span><span class="fu"><a href="https://docs.ropensci.org/rentrez/reference/entrez_search.html" class="external-link">entrez_search</a></span><span class="op">(</span>db<span class="op">=</span><span class="st">"nucleotide"</span>, term<span class="op">=</span><span class="st">"(((txid36200[ORGN] NOT wgs[Keyword]) NOT tsa[Keyword]) AND biomol_genomic[PROP]) AND (cds[Feature key] OR rrna[Feature key])"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Entrez search result with 131 hits (object contains 20 IDs and no web_history object)</span></span>
<span><span class="co">#&gt;  Search term (as translated):  (((txid36200[ORGN] NOT wgs[Keyword]) NOT tsa[Keywo ...</span></span></code></pre></div>
<p>The package recovered every single DNA barcode/gene/product available
for the accession numbers corresponding to <em>Dissostichus
mawsoni</em>, thus, for example, for the accession number ‘LC138011.1’
we obtained both ND1 and ND2 markers as separate records in the
<em>refdb</em> data frame.</p>
<p>In case you want to choose which marker to keep in the final tibble,
set the <em>ask</em> argument to TRUE (default), and, when asked, choose
which marker to download data for:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rec_NCBI</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/download_ncbi.html">download_ncbi</a></span><span class="op">(</span><span class="va">tax</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 43 × 2</span></span></span>
<span><span class="co">#&gt;    Gene_value     n</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> COI            8</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> Dyst           3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> myh6           3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> ND2            3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> tbr1           3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> LDHA           2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> ND4            2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> Ptr            2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> Rhod           2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> TMO-4C4        2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">11</span> ATPase 6       1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">12</span> ATPase 8       1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">13</span> CaM            1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">14</span> CK             1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">15</span> COII           1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">16</span> COIII          1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">17</span> COX1           1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">18</span> Cyt b          1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">19</span> cytb           1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">20</span> Cytb           1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">21</span> EGR1           1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">22</span> EGR2B          1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">23</span> EGR3           1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">24</span> enc1           1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">25</span> ENC1           1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">26</span> G6PDH          1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">27</span> GH             1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">28</span> Mb             1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">29</span> Mll            1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">30</span> MLL            1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">31</span> MLL4           1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">32</span> ND1            1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">33</span> ND3            1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">34</span> ND4L           1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">35</span> ND5            1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">36</span> ND6            1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">37</span> ngb            1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">38</span> RAG1           1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">39</span> rhod           1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">40</span> SEC61          1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">41</span> SH3PX3         1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">42</span> TPI            1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">43</span> zic1           1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Indicate the row numbers corresponding to the CDS and rRNA of interest.</span></span>
<span><span class="co">#&gt;         (0 will return none, other inputs will return all of them)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 63 × 2</span></span></span>
<span><span class="co">#&gt;    Product_value                                                        n</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                                                            <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> 16S ribosomal RNA                                                   33</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> cytochrome oxidase subunit I                                        14</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> cytochrome b                                                         6</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> NADH dehydrogenase subunit 2                                         5</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> hepcidin                                                             4</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> immunoglobulin M heavy chain                                         4</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> 28S ribosomal RNA                                                    3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> 12S ribosomal RNA                                                    2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> AFGP polyprotein precursor                                           2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> chimeric AFGP/trypsinogen-like serine protease precursor             2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">11</span> trypsinogen-like protease 1                                          2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">12</span> 18S ribosomal RNA                                                    1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">13</span> 5S ribosomal RNA                                                     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">14</span> alpha-globin                                                         1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">15</span> antifreeze glycoprotein H1_A1                                        1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">16</span> antifreeze glycoprotein H1_A10                                       1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">17</span> antifreeze glycoprotein H1_A12                                       1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">18</span> antifreeze glycoprotein H1_A2                                        1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">19</span> antifreeze glycoprotein H1_A3                                        1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">20</span> antifreeze glycoprotein H1_A4                                        1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">21</span> antifreeze glycoprotein H1_A5                                        1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">22</span> antifreeze glycoprotein H1_A6                                        1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">23</span> antifreeze glycoprotein H1_A7                                        1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">24</span> antifreeze glycoprotein H1_A8                                        1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">25</span> antifreeze glycoprotein H1_A9                                        1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">26</span> antifreeze glycoprotein H2_A1                                        1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">27</span> antifreeze glycoprotein H2_A2                                        1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">28</span> antifreeze glycoprotein H2_A3                                        1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">29</span> antifreeze glycoprotein H2_A4                                        1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">30</span> antifreeze glycoprotein H2_A7                                        1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">31</span> antifreeze glycoprotein H2_A8                                        1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">32</span> antifreeze glycoprotein/trypsinogen-like protease chimeric H1_C1     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">33</span> antifreeze glycoprotein/trypsinogen-like protease chimeric H1_C2     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">34</span> antifreeze glycoprotein/trypsinogen-like protease chimeric H1_C3     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">35</span> antifreeze glycoprotein/trypsinogen-like protease chimeric H2_C1     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">36</span> antifreeze glycoprotein/trypsinogen-like protease chimeric H2_C2     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">37</span> antifreeze glycoprotein/trypsinogen-like protease chimeric H2_C3     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">38</span> beta-globin                                                          1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">39</span> E3 ubiquitin protein ligase 2                                        1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">40</span> H2_1b                                                                1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">41</span> Mg2+/Mn2+-dependent protein phosphatase 1D                           1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">42</span> MHC class II antigen                                                 1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">43</span> polycystic kidney disease 1                                          1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">44</span> preprotrypsin-like protease                                          1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">45</span> structure specific recognition protein 1                             1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">46</span> translocase of outer mitochondrial membrane 40                       1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">47</span> trypsinogen H1_1c                                                    1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">48</span> trypsinogen H1_1d                                                    1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">49</span> trypsinogen H1_1e                                                    1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">50</span> trypsinogen H1_1f                                                    1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">51</span> trypsinogen H1_1g                                                    1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">52</span> trypsinogen H1_3a1                                                   1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">53</span> trypsinogen H1_3a2                                                   1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">54</span> trypsinogen H1_3a3                                                   1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">55</span> trypsinogen H2_1a                                                    1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">56</span> trypsinogen H2_1c                                                    1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">57</span> trypsinogen H2_1d                                                    1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">58</span> trypsinogen H2_1e                                                    1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">59</span> trypsinogen H2_1f                                                    1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">60</span> trypsinogen H2_1g                                                    1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">61</span> trypsinogen H2_3a1                                                   1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">62</span> trypsinogen H2_3a4                                                   1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">63</span> trypsinogen-like protease 2                                          1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Indicate the row numbers corresponding to the CDS and rRNA of interest.</span></span>
<span><span class="co">#&gt;         (0 will return none, other inputs will return all of them)</span></span></code></pre></div>
<p>The DNA barcodes will be in their 5’-to-3’ direction and of the
correct length defined by the <a href="https://www.insdc.org/submitting-standards/feature-table/#3.4.3" class="external-link">INSDC
location descriptor</a>, as they’ve been extracted from the single fasta
sequence of the same accession number.</p>
<p>The same procedure can be applied using the functions querying the
BOLD database, which have the same naming style. Here, we’re retrieving
the taxonomy information from the BOLD database for the same
species:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># search taxonomic information for a species on BOLD</span></span>
<span><span class="va">tax</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_bold_taxonomy.html">get_bold_taxonomy</a></span><span class="op">(</span><span class="st">"Dissostichus mawsoni"</span><span class="op">)</span></span>
<span><span class="va">tax</span></span>
<span><span class="co">#&gt;              queryName taxid                taxon    rank records</span></span>
<span><span class="co">#&gt; 1 Dissostichus mawsoni 77420 Dissostichus mawsoni species      21</span></span></code></pre></div>
<p>The output is slightly different from the output of <a href="https://matteoce.github.io/barcodeMineR/reference/get_ncbi_taxonomy.html">get_ncbi_taxonomy</a>,
however, it serves the same role. In fact, it can then be used to
download all <em>Dissostichus mawsoni</em> records from the BOLD
database.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rec_BOLD</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/download_bold.html">download_bold</a></span><span class="op">(</span><span class="va">tax</span>, ask <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">rec_BOLD</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 19 × 30</span></span></span>
<span><span class="co">#&gt;    recordID    markerCode DNA_seq phylum class order family genus species source</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;DNA&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> ANGBF23589… COI-5P     <span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #00BB00; font-weight: bold;">CC</span><span style="color: #BBBB00; font-weight: bold;">GG</span><span style="color: #BB0000; font-weight: bold;">A</span>… Chord… Acti… Perc… Notot… Diss… Dissos… BOLD  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> ANGBF23590… COI-5P     <span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #BB0000; font-weight: bold;">AA</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">TT</span>… Chord… Acti… Perc… Notot… Diss… Dissos… BOLD  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> ANTFI127-06 COI-5P     ------… Chord… Acti… Perc… Notot… Diss… Dissos… BOLD  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> ANTFI160-06 COI-5P     --<span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #00BB00; font-weight: bold;">C</span>… Chord… Acti… Perc… Notot… Diss… Dissos… BOLD  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> EATFR006-12 COI-5P     ------… Chord… Acti… Perc… Notot… Diss… Dissos… BOLD  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> EATFR006-12 Rho        <span style="color: #BBBB00; font-weight: bold;">GG</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #00BB00; font-weight: bold;">CC</span>… Chord… Acti… Perc… Notot… Diss… Dissos… BOLD  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> EMRKT082-07 COI-5P     N<span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #BB0000; font-weight: bold;">A</span>… Chord… Acti… Perc… Notot… Diss… Dissos… BOLD  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> FNZ726-06   COI-5P     -<span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #BB0000; font-weight: bold;">A</span>… Chord… Acti… Perc… Notot… Diss… Dissos… BOLD  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> FNZ727-06   COI-5P     -<span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #BB0000; font-weight: bold;">A</span>… Chord… Acti… Perc… Notot… Diss… Dissos… BOLD  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> FNZ728-06   COI-5P     -<span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #BB0000; font-weight: bold;">A</span>… Chord… Acti… Perc… Notot… Diss… Dissos… BOLD  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">11</span> FNZ969-07   COI-5P     <span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #BB0000; font-weight: bold;">A</span>… Chord… Acti… Perc… Notot… Diss… Dissos… BOLD  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">12</span> FNZC284-09  COI-5P     -<span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #BB0000; font-weight: bold;">A</span>… Chord… Acti… Perc… Notot… Diss… Dissos… BOLD  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">13</span> GBMNA17173… COI-5P     <span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #BBBB00; font-weight: bold;">GG</span><span style="color: #00BB00; font-weight: bold;">CC</span>… Chord… Acti… Perc… Notot… Diss… Dissos… BOLD  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">14</span> GBMNB8127-… COI-5P     <span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #BB0000; font-weight: bold;">A</span>… Chord… Acti… Perc… Notot… Diss… Dissos… BOLD  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">15</span> GBMNF47574… COI-5P     <span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #00BB00; font-weight: bold;">CC</span><span style="color: #0000BB; font-weight: bold;">T</span>… Chord… Acti… Perc… Notot… Diss… Dissos… BOLD  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">16</span> OCARH832-12 COI-5P     --<span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #00BB00; font-weight: bold;">C</span>… Chord… Acti… Perc… Notot… Diss… Dissos… BOLD  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">17</span> OCARH904-12 COI-5P     --<span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #00BB00; font-weight: bold;">C</span>… Chord… Acti… Perc… Notot… Diss… Dissos… BOLD  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">18</span> OCARH973-12 COI-5P     --<span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #00BB00; font-weight: bold;">C</span>… Chord… Acti… Perc… Notot… Diss… Dissos… BOLD  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">19</span> OCARH981-12 COI-5P     --<span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #00BB00; font-weight: bold;">C</span>… Chord… Acti… Perc… Notot… Diss… Dissos… BOLD  </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 20 more variables: lat &lt;dbl&gt;, lon &lt;dbl&gt;, lengthGene &lt;int&gt;, sampleID &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   QueryName &lt;chr&gt;, identified_by &lt;chr&gt;, taxNotes &lt;lgl&gt;, db_xref &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   sourceID &lt;chr&gt;, NCBI_ID &lt;chr&gt;, institutionStoring &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   collected_by &lt;chr&gt;, collection_date &lt;lgl&gt;, altitude &lt;int&gt;, depth &lt;int&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   country &lt;chr&gt;, directionPrimers &lt;chr&gt;, lengthSource &lt;int&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   PCR_primers &lt;chr&gt;, note &lt;chr&gt;</span></span></span></code></pre></div>
<p>We have retrieved 19 records from the BOLD database, corresponding to
18 different process-IDs. BOLD hosts other markers, but the vast
majority of records correspond to Cytochrome c oxidase I sequences.</p>
<p>Searching species names, or taxa in general, using the <a href="https://matteoce.github.io/barcodeMineR/reference/index.html#taxonomy">taxonomy
functions</a> follows a set of rules that is described in the vignette
“<a href="https://matteoce.github.io/barcodeMineR/articles/taxonomy_barcodeMineR.html">Searching
taxonomy</a>”.</p>
<div class="section level2">
<h2 id="merging-data-from-different-repositories">Merging data from different repositories<a class="anchor" aria-label="anchor" href="#merging-data-from-different-repositories"></a>
</h2>
<p>We can now merge both <em>refdb</em> data frames to obtain a single
one. The default behaviour of the next function will check which records
have been mined from the other repository and return only the original
record. If needed, this behaviour can be switched off by setting the
argument <em>resolve.conflicts</em> to FALSE:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">total</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mergeBarcodeOres.html">mergeBarcodeOres</a></span><span class="op">(</span><span class="va">rec_NCBI</span>, <span class="va">rec_BOLD</span><span class="op">)</span></span>
<span><span class="co">#&gt; '5' records from BOLD were mined from the NCBI.</span></span>
<span><span class="co">#&gt; If they are already represented by the NCBI barcodeOre they will be removed to avoid duplicates.</span></span>
<span><span class="co">#&gt; '3' records from NCBI were mined from BOLD.</span></span>
<span><span class="co">#&gt; If they are already represented by the BOLD barcodeOre they will be removed to avoid duplicates.</span></span>
<span><span class="co">#&gt; Duplicated records obtained from the BOLD will be kept.</span></span></code></pre></div>
<p>As we can see from the printed messages, some records were actually
mined versions of other records obtained from the other repository. They
have been removed after checking that the original version was
downloaded from the opposite repository.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">total</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 203 × 30</span></span></span>
<span><span class="co">#&gt;    recordID   markerCode DNA_seq  phylum class order family genus species source</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;DNA&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> ON000293.1 COX1       <span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #00BB00; font-weight: bold;">CC</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #BBBB00; font-weight: bold;">G</span>… Chord… Acti… Perc… Notot… Diss… Dissos… NCBI  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> MK843765.1 COI        <span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #BB0000; font-weight: bold;">A</span><span style="color: #00BB00; font-weight: bold;">C</span>… Chord… Acti… Perc… Notot… Diss… Dissos… NCBI  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> DQ498816.1 Cytb       <span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #00BB00; font-weight: bold;">CC</span><span style="color: #BB0000; font-weight: bold;">A</span><span style="color: #00BB00; font-weight: bold;">CCC</span>… Chord… Acti… Perc… Notot… Diss… Dissos… NCBI  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> DQ498794.1 Rhod       <span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #00BB00; font-weight: bold;">CC</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #BB0000; font-weight: bold;">A</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #BB0000; font-weight: bold;">A</span>… Chord… Acti… Perc… Notot… Diss… Dissos… NCBI  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> MK500763.1 enc1       <span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #BB0000; font-weight: bold;">A</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #BBBB00; font-weight: bold;">G</span>… Chord… Acti… Perc… Notot… Diss… Dissos… NCBI  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> MG729451.1 COI        <span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #BB0000; font-weight: bold;">AA</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">TT</span><span style="color: #BB0000; font-weight: bold;">A</span>… Chord… Acti… Perc… Notot… Diss… Dissos… NCBI  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> KY656477.1 COI        <span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #00BB00; font-weight: bold;">CC</span><span style="color: #BBBB00; font-weight: bold;">GG</span><span style="color: #BB0000; font-weight: bold;">AA</span>… Chord… Acti… Perc… Notot… Diss… Dissos… NCBI  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> LC138011.1 ND1        <span style="color: #BB0000; font-weight: bold;">A</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">TTT</span>… Chord… Acti… Perc… Notot… Diss… Dissos… NCBI  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> LC138011.1 ND2        <span style="color: #BB0000; font-weight: bold;">A</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #BB0000; font-weight: bold;">A</span><span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #00BB00; font-weight: bold;">CC</span>… Chord… Acti… Perc… Notot… Diss… Dissos… NCBI  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> LC138011.1 COI        <span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #BBBB00; font-weight: bold;">GG</span><span style="color: #00BB00; font-weight: bold;">CC</span><span style="color: #BB0000; font-weight: bold;">A</span>… Chord… Acti… Perc… Notot… Diss… Dissos… NCBI  </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 193 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 20 more variables: lat &lt;dbl&gt;, lon &lt;dbl&gt;, lengthGene &lt;int&gt;, sampleID &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   QueryName &lt;chr&gt;, identified_by &lt;chr&gt;, taxNotes &lt;lgl&gt;, db_xref &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   sourceID &lt;chr&gt;, NCBI_ID &lt;chr&gt;, institutionStoring &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   collected_by &lt;chr&gt;, collection_date &lt;chr&gt;, altitude &lt;int&gt;, depth &lt;int&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   country &lt;chr&gt;, directionPrimers &lt;chr&gt;, lengthSource &lt;int&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   PCR_primers &lt;chr&gt;, note &lt;chr&gt;</span></span></span></code></pre></div>
<p>This is it, now we have all DNA barcodes available on the major
online repositories for the species <em>Dissostichus mawsoni</em>.</p>
<p>In case we want to add additional, private sequences to the mined
results, we can use the function <a href="https://matteoce.github.io/barcodeMineR/reference/loadBarcodeOre.html">loadBarcodeOre</a>
and specify either the paths to a tsv file and a fasta file including
our new sequences, or to a data frame and a <em>DNAStringSet</em> object
with the same, to output a <em>refdb</em> data frame formatted as those
we merged earlier.</p>
<p>In order to do that, the tsv file must include the same fields
(<em>i.e.</em> column names) included in the example dataset object
<em>example_record</em>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load example datasets from barcodeMineR...</span></span>
<span><span class="va">rec</span> <span class="op">&lt;-</span> <span class="fu">barcodeMineR</span><span class="fu">::</span><span class="va"><a href="../reference/example_record.html">example_record</a></span></span>
<span><span class="va">seq</span> <span class="op">&lt;-</span> <span class="fu">barcodeMineR</span><span class="fu">::</span><span class="va"><a href="../reference/example_sequence.html">example_sequence</a></span></span>
<span></span>
<span><span class="co">#... or include the path to your files</span></span>
<span><span class="va">rec</span> <span class="op">&lt;-</span> <span class="st">"~/home/my/path/data.tsv"</span></span>
<span><span class="va">seq</span> <span class="op">&lt;-</span> <span class="st">"~/home/my/path/data.fasta"</span></span>
<span></span>
<span><span class="va">new_records</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/loadBarcodeOre.html">loadBarcodeOre</a></span><span class="op">(</span><span class="va">rec</span>, <span class="va">seq</span><span class="op">)</span></span>
<span><span class="va">new_records</span></span></code></pre></div>
<pre><code><span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 30</span></span></span>
<span><span class="co">#&gt;   recordID markerCode DNA_seq     phylum class order family genus species source</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;DNA&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> SEQ_01   COI        <span style="color: #BB0000; font-weight: bold;">AAA</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #BB0000; font-weight: bold;">AAA</span><span style="color: #BBBB00; font-weight: bold;">G</span>… Chord… Acti… Perc… Notot… Diss… Dissos… ACRON…</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 20 more variables: lat &lt;dbl&gt;, lon &lt;dbl&gt;, lengthGene &lt;int&gt;, sampleID &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   QueryName &lt;chr&gt;, identified_by &lt;chr&gt;, taxNotes &lt;chr&gt;, db_xref &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   sourceID &lt;chr&gt;, NCBI_ID &lt;chr&gt;, institutionStoring &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   collected_by &lt;chr&gt;, collection_date &lt;chr&gt;, altitude &lt;chr&gt;, depth &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   country &lt;chr&gt;, directionPrimers &lt;chr&gt;, lengthSource &lt;int&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   PCR_primers &lt;chr&gt;, note &lt;chr&gt;</span></span></span></code></pre>
<p>We can merge these records with the ones we mined online and examine
all results:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">total</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mergeBarcodeOres.html">mergeBarcodeOres</a></span><span class="op">(</span><span class="va">total</span>, <span class="va">new_records</span><span class="op">)</span></span>
<span><span class="co">#&gt; No records were obtained from both the NCBI and BOLD</span></span>
<span><span class="va">total</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 204 × 30</span></span></span>
<span><span class="co">#&gt;    recordID   markerCode DNA_seq  phylum class order family genus species source</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;DNA&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> ON000293.1 COX1       <span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #00BB00; font-weight: bold;">CC</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #BBBB00; font-weight: bold;">G</span>… Chord… Acti… Perc… Notot… Diss… Dissos… NCBI  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> MK843765.1 COI        <span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #BB0000; font-weight: bold;">A</span><span style="color: #00BB00; font-weight: bold;">C</span>… Chord… Acti… Perc… Notot… Diss… Dissos… NCBI  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> DQ498816.1 Cytb       <span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #00BB00; font-weight: bold;">CC</span><span style="color: #BB0000; font-weight: bold;">A</span><span style="color: #00BB00; font-weight: bold;">CCC</span>… Chord… Acti… Perc… Notot… Diss… Dissos… NCBI  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> DQ498794.1 Rhod       <span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #00BB00; font-weight: bold;">CC</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #BB0000; font-weight: bold;">A</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #BB0000; font-weight: bold;">A</span>… Chord… Acti… Perc… Notot… Diss… Dissos… NCBI  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> MK500763.1 enc1       <span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #BB0000; font-weight: bold;">A</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #BBBB00; font-weight: bold;">G</span>… Chord… Acti… Perc… Notot… Diss… Dissos… NCBI  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> MG729451.1 COI        <span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #BB0000; font-weight: bold;">AA</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">TT</span><span style="color: #BB0000; font-weight: bold;">A</span>… Chord… Acti… Perc… Notot… Diss… Dissos… NCBI  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> KY656477.1 COI        <span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #00BB00; font-weight: bold;">CC</span><span style="color: #BBBB00; font-weight: bold;">GG</span><span style="color: #BB0000; font-weight: bold;">AA</span>… Chord… Acti… Perc… Notot… Diss… Dissos… NCBI  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> LC138011.1 ND1        <span style="color: #BB0000; font-weight: bold;">A</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #00BB00; font-weight: bold;">C</span><span style="color: #0000BB; font-weight: bold;">TTT</span>… Chord… Acti… Perc… Notot… Diss… Dissos… NCBI  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> LC138011.1 ND2        <span style="color: #BB0000; font-weight: bold;">A</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #BB0000; font-weight: bold;">A</span><span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #00BB00; font-weight: bold;">CC</span>… Chord… Acti… Perc… Notot… Diss… Dissos… NCBI  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> LC138011.1 COI        <span style="color: #BBBB00; font-weight: bold;">G</span><span style="color: #0000BB; font-weight: bold;">T</span><span style="color: #BBBB00; font-weight: bold;">GG</span><span style="color: #00BB00; font-weight: bold;">CC</span><span style="color: #BB0000; font-weight: bold;">A</span>… Chord… Acti… Perc… Notot… Diss… Dissos… NCBI  </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 194 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 20 more variables: lat &lt;dbl&gt;, lon &lt;dbl&gt;, lengthGene &lt;int&gt;, sampleID &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   QueryName &lt;chr&gt;, identified_by &lt;chr&gt;, taxNotes &lt;chr&gt;, db_xref &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   sourceID &lt;chr&gt;, NCBI_ID &lt;chr&gt;, institutionStoring &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   collected_by &lt;chr&gt;, collection_date &lt;chr&gt;, altitude &lt;chr&gt;, depth &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   country &lt;chr&gt;, directionPrimers &lt;chr&gt;, lengthSource &lt;int&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   PCR_primers &lt;chr&gt;, note &lt;chr&gt;</span></span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="analyzing-the-database">Analyzing the database<a class="anchor" aria-label="anchor" href="#analyzing-the-database"></a>
</h2>
<p>Having a pre-formatted <a href="https://fkeck.github.io/refdb/index.html" class="external-link">refdb</a> data frame
allows to use the functions of that package, including filtering
operations, quality checks, examination of taxonomic conflicts and
plotting functions. For example, if we were to search for taxonomic
conflicts, we could use the appropriate function from the <em>refdb</em>
package:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://fkeck.github.io/refdb/" class="external-link">refdb</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://fkeck.github.io/refdb/reference/refdb_check_tax_conflict.html" class="external-link">refdb_check_tax_conflict</a></span><span class="op">(</span><span class="va">total</span><span class="op">)</span></span>
<span><span class="co">#&gt; $phylum</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 0 × 2</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 2 variables: Taxon &lt;chr&gt;, Taxonomy &lt;???&gt;</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $class</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 0 × 2</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 2 variables: Taxon &lt;chr&gt;, Taxonomy &lt;???&gt;</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $order</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 2</span></span></span>
<span><span class="co">#&gt;   Taxon       Taxonomy                 </span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Perciformes Chordata &gt; Actinopteri   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Perciformes Chordata &gt; Actinopterygii</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $family</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 2</span></span></span>
<span><span class="co">#&gt;   Taxon         Taxonomy                               </span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                                  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Nototheniidae Chordata &gt; Actinopteri &gt; Perciformes   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Nototheniidae Chordata &gt; Actinopterygii &gt; Perciformes</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $genus</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 2</span></span></span>
<span><span class="co">#&gt;   Taxon        Taxonomy                                               </span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                                                  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Dissostichus Chordata &gt; Actinopteri &gt; Perciformes &gt; Nototheniidae   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Dissostichus Chordata &gt; Actinopterygii &gt; Perciformes &gt; Nototheniidae</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $species</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 2</span></span></span>
<span><span class="co">#&gt;   Taxon                Taxonomy                                                 </span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                                                    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Dissostichus mawsoni Chordata &gt; Actinopteri &gt; Perciformes &gt; Nototheniidae &gt; D…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Dissostichus mawsoni Chordata &gt; Actinopterygii &gt; Perciformes &gt; Nototheniidae …</span></span></code></pre></div>
<p>As observed from the output, it appears that the nomenclature for
ray-finned fishes is different in the BOLD and NCBI taxonomies
(Actinopteri vs Actinopterygii).</p>
<p>We can also filter records based on the length of the sequences.
Although we retain this information in the lengthGene field, the
<em>refdb</em> package directly calculates it using the <a href="https://fkeck.github.io/refdb/reference/refdb_filter_seq_length.html" class="external-link">refdb_filter_seq_length</a>
function. Let’s say we want to know how many records correspond to a
maximum of 658 base pairs (bp) in length:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">total_filt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://fkeck.github.io/refdb/reference/refdb_filter_seq_length.html" class="external-link">refdb_filter_seq_length</a></span><span class="op">(</span><span class="va">total</span>, max_len <span class="op">=</span> <span class="fl">658</span><span class="op">)</span></span></code></pre></div>
<p>A total of 99 records include sequences up to 658 bp in length.
However, many records results much shorter than this length:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://fkeck.github.io/refdb/reference/refdb_plot_seqlen_hist.html" class="external-link">refdb_plot_seqlen_hist</a></span><span class="op">(</span><span class="va">total_filt</span><span class="op">)</span></span></code></pre></div>
<p><img src="intro_barcodeMineR_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
<p>In fact, many records have been extracted from complete mitochondrial
genomes, and this can be readily inspected by using one of the plotting
functions of the <em>barcodeMineR</em> package:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_length.html">plot_length</a></span><span class="op">(</span><span class="va">total_filt</span><span class="op">)</span></span></code></pre></div>
<p><img src="intro_barcodeMineR_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
<p>As shown, the original length of the fasta sequence of some records
is much longer than the final result obtained with the
<em>barcodeMineR</em> package.</p>
<p>Not much information was retrieved from each record, but we can see
how many records include the primers used to amplify their sequence with
another plotting function of the <em>barcodeMineR</em> package, credit
by <a href="https://journals.plos.org/plosone/article/file?id=10.1371/journal.pone.0125635&amp;type=printable" class="external-link">Gwiazdowski
et al. (2015)</a>:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_primers.html">plot_primers</a></span><span class="op">(</span><span class="va">total_filt</span><span class="op">)</span></span></code></pre></div>
<p><img src="intro_barcodeMineR_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
<p>We can also map the records with available geographical coordinates
using the <em>refdb</em> package:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://fkeck.github.io/refdb/reference/refdb_plot_map.html" class="external-link">refdb_plot_map</a></span><span class="op">(</span><span class="va">total</span><span class="op">)</span></span></code></pre></div>
<div class="leaflet html-widget html-fill-item" id="htmlwidget-ac96cb3ee4656e2e9ec3" style="width:700px;height:432.632880098888px;"></div>
<script type="application/json" data-for="htmlwidget-ac96cb3ee4656e2e9ec3">{"x":{"options":{"crs":{"crsClass":"L.CRS.EPSG3857","code":null,"proj4def":null,"projectedBounds":null,"options":{}}},"calls":[{"method":"addTiles","args":["https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png",null,null,{"minZoom":0,"maxZoom":18,"tileSize":256,"subdomains":"abc","errorTileUrl":"","tms":false,"noWrap":false,"zoomOffset":0,"zoomReverse":false,"opacity":1,"zIndex":1,"detectRetina":false,"attribution":"&copy; <a href=\"https://openstreetmap.org/copyright/\">OpenStreetMap<\/a>,  <a href=\"https://opendatacommons.org/licenses/odbl/\">ODbL<\/a>"}]},{"method":"addCircleMarkers","args":[[null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,-61.22,-61.13,-77.1699,-77.1699,null,null,null,null,null,null,null,null,null,null,-74.6789],[null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,-55.67,-55.92,-177.565,-177.565,null,null,null,null,null,null,null,null,null,null,164.2315],8,null,null,{"interactive":true,"className":"","stroke":false,"color":"#03F","weight":5,"opacity":0.5,"fill":true,"fillColor":"red","fillOpacity":1},{"showCoverageOnHover":true,"zoomToBoundsOnClick":true,"spiderfyOnMaxZoom":true,"removeOutsideVisibleBounds":true,"spiderLegPolylineOptions":{"weight":1.5,"color":"#222","opacity":0.5},"freezeAtZoom":false},null,["<h3>ID: ON000293.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>635","<h3>ID: MK843765.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>646","<h3>ID: DQ498816.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>541","<h3>ID: DQ498794.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>758","<h3>ID: MK500763.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>732","<h3>ID: MG729451.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>539","<h3>ID: KY656477.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>629","<h3>ID: LC138011.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>975","<h3>ID: LC138011.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>1046","<h3>ID: LC138011.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>1551","<h3>ID: LC138011.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>691","<h3>ID: LC138011.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>168","<h3>ID: LC138011.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>696","<h3>ID: LC138011.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>785","<h3>ID: LC138011.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>349","<h3>ID: LC138011.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>297","<h3>ID: LC138011.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>1381","<h3>ID: LC138011.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>1839","<h3>ID: LC138011.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>1141","<h3>ID: LC138011.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>525","<h3>ID: JN640902.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>652","<h3>ID: EF535029.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>13","<h3>ID: DQ184498.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>1047","<h3>ID: AY517803.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>456","<h3>ID: AF137119.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>120","<h3>ID: AF134618.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>38","<h3>ID: AF137249.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>810","<h3>ID: AF137152.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>114","<h3>ID: AY256561.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>1047","<h3>ID: EF088427.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>178","<h3>ID: EF088426.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>567","<h3>ID: EF088425.1_1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>47","<h3>ID: EF088425.1_2<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>208","<h3>ID: EF088424.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>511","<h3>ID: EF088423.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>660","<h3>ID: EF088422.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>297","<h3>ID: EF088421.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>51","<h3>ID: EF088420.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>726","<h3>ID: EF088419.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>64","<h3>ID: EF088418.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>106","<h3>ID: EF088417.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>58","<h3>ID: KJ721665.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>636","<h3>ID: KJ721605.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>801","<h3>ID: KC442215.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>1497","<h3>ID: KC442188.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>903","<h3>ID: KC442148.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>861","<h3>ID: KC442115.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>855","<h3>ID: KF312023.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>618","<h3>ID: KC250350.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>319","<h3>ID: JF264603.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>642","<h3>ID: JF264602.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>623","<h3>ID: JF264567.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>702","<h3>ID: JF264566.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>702","<h3>ID: JF264529.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>705","<h3>ID: JF264528.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>705","<h3>ID: JN187040.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>720","<h3>ID: JN186999.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>636","<h3>ID: JN186958.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>699","<h3>ID: JN186822.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>831","<h3>ID: FM204871.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>444","<h3>ID: MZ170973.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>799","<h3>ID: MZ170972.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>799","<h3>ID: MZ170971.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>799","<h3>ID: MZ170970.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>799","<h3>ID: MZ170969.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>799","<h3>ID: MZ170968.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>648","<h3>ID: MZ170967.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>648","<h3>ID: MZ170966.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>648","<h3>ID: MZ170965.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>648","<h3>ID: MW684761.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>619","<h3>ID: MW684760.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>619","<h3>ID: MW684759.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>619","<h3>ID: MW684758.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>619","<h3>ID: MW684757.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>619","<h3>ID: MW684756.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>619","<h3>ID: MW684755.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>619","<h3>ID: MW684754.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>619","<h3>ID: MW684753.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>619","<h3>ID: MW684752.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>619","<h3>ID: MW684751.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>619","<h3>ID: MW684750.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>619","<h3>ID: MW684749.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>619","<h3>ID: MW684748.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>619","<h3>ID: MW684747.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>619","<h3>ID: MW684746.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>619","<h3>ID: MW684745.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>619","<h3>ID: MW684744.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>619","<h3>ID: MW684743.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>619","<h3>ID: MW684742.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>619","<h3>ID: MW684741.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>619","<h3>ID: MW684740.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>619","<h3>ID: MW684739.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>619","<h3>ID: MW684738.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>619","<h3>ID: MW684737.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>619","<h3>ID: MW684736.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>619","<h3>ID: MW684735.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>619","<h3>ID: MW684734.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>619","<h3>ID: MW684733.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>685","<h3>ID: MW684732.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>685","<h3>ID: MW684731.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>685","<h3>ID: MW684730.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>685","<h3>ID: MW684729.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>685","<h3>ID: MW684728.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>685","<h3>ID: MW684727.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>685","<h3>ID: MW684726.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>685","<h3>ID: MW684725.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>685","<h3>ID: MW684724.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>685","<h3>ID: MW684723.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>685","<h3>ID: MW684722.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>685","<h3>ID: MW684721.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>685","<h3>ID: MW684720.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>685","<h3>ID: Z32726.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>542","<h3>ID: Z32707.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>373","<h3>ID: LC138011.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>948","<h3>ID: LC138011.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>1694","<h3>ID: JN034583.1_1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>100","<h3>ID: JN034583.1_2<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>316","<h3>ID: JN034583.1_3<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>391","<h3>ID: JN034583.1_4<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>77","<h3>ID: HQ447060.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>729","<h3>ID: HQ447060.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>729","<h3>ID: HQ447060.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>726","<h3>ID: HQ447060.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>726","<h3>ID: HQ447060.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>729","<h3>ID: HQ447060.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>717","<h3>ID: HQ447060.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>729","<h3>ID: HQ447060.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>738","<h3>ID: HQ447060.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>744","<h3>ID: HQ447060.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>1754","<h3>ID: HQ447060.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>675","<h3>ID: HQ447060.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>2529","<h3>ID: HQ447060.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>1362","<h3>ID: HQ447060.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>3681","<h3>ID: HQ447060.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>1761","<h3>ID: HQ447060.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>2514","<h3>ID: HQ447060.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>2064","<h3>ID: HQ447060.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>2349","<h3>ID: HQ447060.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>1956","<h3>ID: HQ447059.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>726","<h3>ID: HQ447059.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>729","<h3>ID: HQ447059.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>729","<h3>ID: HQ447059.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>729","<h3>ID: HQ447059.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>726","<h3>ID: HQ447059.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>738","<h3>ID: HQ447059.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>750","<h3>ID: HQ447059.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>738","<h3>ID: HQ447059.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>744","<h3>ID: HQ447059.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>696","<h3>ID: HQ447059.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>1944","<h3>ID: HQ447059.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>2189","<h3>ID: HQ447059.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>1932","<h3>ID: HQ447059.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>2187","<h3>ID: HQ447059.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>1770","<h3>ID: HQ447059.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>1770","<h3>ID: HQ447059.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>1674","<h3>ID: HQ447059.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>1344","<h3>ID: HQ447059.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>2892","<h3>ID: HQ447059.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>2457","<h3>ID: HQ447059.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>2124","<h3>ID: HQ447059.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>2001","<h3>ID: HQ447059.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>1529","<h3>ID: HQ447059.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>1916","<h3>ID: HQ447059.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>1008","<h3>ID: AF134961.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>188","<h3>ID: AF134321.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>2030","<h3>ID: AF134320.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>40","<h3>ID: JF264491.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>1099","<h3>ID: JF264490.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>1099","<h3>ID: JQ688869.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>203","<h3>ID: JQ688816.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>565","<h3>ID: JQ688764.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>868","<h3>ID: JQ688715.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>436","<h3>ID: FJ896404.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>432","<h3>ID: FJ896404.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>444","<h3>ID: AY520111.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>1692","<h3>ID: AY520110.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>1692","<h3>ID: AF518188.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>1579","<h3>ID: GU018108.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>528","<h3>ID: U58944.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>2169","<h3>ID: U58835.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>750","<h3>ID: U43149.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>1089","<h3>ID: EU221603.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>261","<h3>ID: EU221599.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>261","<h3>ID: EU221596.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>267","<h3>ID: EU221590.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>273","<h3>ID: DQ478734.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>86","<h3>ID: AY926497.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>762","<h3>ID: U87436.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>346","<h3>ID: U87423.1<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>433","<h3>ID: ANTFI127-06<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>654","<h3>ID: ANTFI160-06<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>654","<h3>ID: EATFR006-12<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>653","<h3>ID: EATFR006-12<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>777","<h3>ID: EMRKT082-07<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>652","<h3>ID: FNZ726-06<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>652","<h3>ID: FNZ727-06<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>652","<h3>ID: FNZ728-06<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>652","<h3>ID: FNZ969-07<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>652","<h3>ID: FNZC284-09<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>652","<h3>ID: OCARH832-12<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>657","<h3>ID: OCARH904-12<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>657","<h3>ID: OCARH973-12<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>657","<h3>ID: OCARH981-12<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>657","<h3>ID: SEQ_01<\/h3><b>Species: <\/b>Dissostichus mawsoni<br><b>Sequence length: <\/b>658"],null,null,{"interactive":false,"permanent":false,"direction":"auto","opacity":1,"offset":[0,0],"textsize":"10px","textOnly":false,"className":"","sticky":true},null]}],"limits":{"lat":[-77.1699,-61.13],"lng":[-177.565,164.2315]}},"evals":[],"jsHooks":[]}</script><p>Only a few records have geographical coordinates in the final merged
table.</p>
</div>
<div class="section level2">
<h2 id="exporting-the-results">Exporting the results<a class="anchor" aria-label="anchor" href="#exporting-the-results"></a>
</h2>
<p>Once one or more <em>refdb</em> objects has been obtained, these can
be exported as any data frame. As an example, here it is shown how to
export one as a TSV file, using the package <em>readr</em>:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html" class="external-link">write_delim</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">total</span>, file <span class="op">=</span> <span class="st">"~/my/path/to/file.tsv"</span>, delim <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span></span></code></pre></div>
<p>Likewise, if this file needs to be loaded as a simple data.frame
object, the same method can be applied. However, in case this object
needs to be loaded as a <em>refdb</em> data frame, then the fields set
by the <em>barcodeMineR</em> functions need to be preserved in a
<em>yaml</em> file designed specifically for its outputs. For this
reason, when saving a <em>refdb</em> object obtained with the
<em>barcodeMineR</em> functions, a <em>yaml</em> file must be stored
alongside it. This can be done using the <em>refdb</em> package:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># save the fields settings in a yml file format</span></span>
<span><span class="fu"><a href="https://fkeck.github.io/refdb/reference/refdb_write_fields.html" class="external-link">refdb_write_fields</a></span><span class="op">(</span><span class="va">total</span>, <span class="st">"~/my/path/to/fields.yml"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># load the tsv saved before and set the fields </span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_delim</a></span><span class="op">(</span>file<span class="op">=</span><span class="st">"~/my/path/to/file.tsv"</span>, delim <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span></span>
<span><span class="va">ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://fkeck.github.io/refdb/reference/refdb_set_fields.html" class="external-link">refdb_set_fields</a></span><span class="op">(</span><span class="va">df</span>, config_yaml <span class="op">=</span> <span class="st">"~/my/path/to/fields.yml"</span><span class="op">)</span></span></code></pre></div>
<p>This is the basic usage of the <strong>barcodeMineR</strong> package.
Now you can check how to speed up the recovery of records from the NCBI
database by reading this <a href="https://matteoce.github.io/barcodeMineR/articles/api_rate_barcodeMineR.html">vignette</a>.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Matteo Cecchetto, Stefano Schiaparelli.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
